### PR TITLE
mortars now automatically link and save deployed binoculars

### PR DESCRIPTION
**Pull request by MacBlaze1**
## About The Pull Request
Mortars now automatically link to saved binoculars and keep track of them in item state as well as structure state

## Why It's Good For The Game

Massive QOL, plus bug fix, plus I need street credit with the engineers

## Changelog

:cl:

qol: mortars now automatically set save binoculars to them, linking with saved binoculars, and keeping track of them, no more need to use a binocular on a mortar before firing.

/:cl:

### CODE REVIEW

**Comment by Losenis** [comment]
Can you extend the bullet hitting to all deployables? mfs hitting my floodlights and chem dispensers on aim mode not even noticing because they're hyperfixated on the xeno 3 screens downrange

**Comment by MacBlaze1** [comment]
> Can you extend the bullet hitting to all deployables? mfs hitting my floodlights and chem dispensers on aim mode not even noticing because they're hyperfixated on the xeno 3 screens downrange

done

**Comment by zzurge** [comment]
> > Can you extend the bullet hitting to all deployables? mfs hitting my floodlights and chem dispensers on aim mode not even noticing because they're hyperfixated on the xeno 3 screens downrange
> 
> done

BASED I KNEEL

File: code/game/objects/machinery/deployable.dm [hunk 1]
```diff
@@ -2,7 +2,7 @@
 	flags_atom = PREVENT_CONTENTS_EXPLOSION
 	hud_possible = list(MACHINE_HEALTH_HUD)
 	obj_flags = CAN_BE_HIT
-	flags_pass = PASSAIR
+	flags_pass = PASSABLE
```
Comment (TiviPlus) [inline code/game/objects/machinery/deployable.dm:5]: ??

File: code/game/objects/machinery/mortar.dm [hunk 2]
```diff
@@ -72,9 +72,18 @@
...
 	var/static/list/id_by_type = list()
 
+	var/list/linked_binoculars = list()
+	var/obj/item/mortar_kit/linked_mortar
+
 /obj/machinery/deployable/mortar/Initialize(mapload, _internal_item, deployer)
 	. = ..()
 
+	RegisterSignal(src, COMSIG_ITEM_UNDEPLOY, PROC_REF(handle_undeploy_references))
+	linked_mortar = get_internal_item()
```
Comment (TiviPlus) [inline code/game/objects/machinery/mortar.dm:82]: Why does the linked_mortar var exist if you can just get_internal_item

File: code/game/objects/machinery/mortar.dm [hunk 3]
```diff
@@ -313,6 +324,13 @@
 	ai_targeter.forceMove(src.loc)
 	ai_targeter = null
 
+/obj/machinery/deployable/mortar/proc/handle_undeploy_references()
+	linked_mortar = get_internal_item()
+	LAZYINITLIST(linked_mortar.linked_binoculars)
+	LAZYCLEARLIST(linked_mortar.linked_binoculars)
+	for(var/binoc in src.linked_binoculars)
+		LAZYADD(linked_mortar.linked_binoculars, binoc)
```
Comment (TiviPlus) [inline code/game/objects/machinery/mortar.dm:332]: why not just set the vars to each other or use .Copy()?

File: code/game/objects/machinery/mortar.dm [hunk 4]
```diff
@@ -401,6 +419,7 @@
 	var/deployable_item = /obj/machinery/deployable/mortar
 	resistance_flags = RESIST_ALL
 	w_class = WEIGHT_CLASS_BULKY //No dumping this in most backpacks. Carry it, fatso
+	var/list/linked_binoculars
```
Comment (TiviPlus) [inline code/game/objects/machinery/mortar.dm:422]: Is this used anywhere?
either way the identical names make it confusing which is which + no autodoc

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/objects/machinery/mortar.dm [hunk 5]
```diff
@@ -239,8 +248,10 @@
 	var/obj/item/binoculars/tactical/binocs = I
 	playsound(src, 'sound/effects/binoctarget.ogg', 35)
 	if(binocs.set_mortar(src))
+		LAZYADD(linked_binoculars, binocs)
 		balloon_alert(user, "linked")
 		return
+	LAZYREMOVE(linked_binoculars, binocs)
```
[the remark this answers is not in the document]
Comment (rhit-summermt) [reply code/game/objects/machinery/mortar.dm:254]: Ill make set mortar handle it

File: code/game/objects/machinery/deployable.dm (hunk 1 above)
Comment (MacBlaze1) [reply code/game/objects/machinery/deployable.dm]: removed, Ill make it in another PR

File: code/game/objects/machinery/mortar.dm [hunk 6]
```diff
@@ -72,9 +72,18 @@
 	// used for keeping track of different mortars and their types for cams
 	var/static/list/id_by_type = list()
 
+	var/list/linked_binoculars = list()
+	var/obj/item/mortar_kit/linked_mortar
+
```
[the remark this answers is not in the document]
Comment (MacBlaze1) [reply code/game/objects/machinery/mortar.dm]: autodocked I think

File: code/game/objects/machinery/mortar.dm (hunk 2 above)
Comment (MacBlaze1) [reply code/game/objects/machinery/mortar.dm]: using internal item from now on

File: code/game/objects/machinery/mortar.dm [hunk 7]
```diff
@@ -72,9 +72,18 @@
...
 /obj/machinery/deployable/mortar/Initialize(mapload, _internal_item, deployer)
 	. = ..()
 
+	RegisterSignal(src, COMSIG_ITEM_UNDEPLOY, PROC_REF(handle_undeploy_references))
+	linked_mortar = get_internal_item()
+	LAZYINITLIST(linked_binoculars)
+	for (var/obj/item/binoculars/tactical/binoc in linked_mortar.linked_binoculars)
+		LAZYADD(linked_binoculars, binoc)
+		binoc.set_mortar(src)
```
[the remark this answers is not in the document]
Comment (MacBlaze1) [reply code/game/objects/machinery/mortar.dm]: set mortar now handles this

File: code/game/objects/machinery/mortar.dm (hunk 5 above)
[the remark this answers is not in the document]
Comment (MacBlaze1) [reply code/game/objects/machinery/mortar.dm]: it does now

File: code/game/objects/machinery/mortar.dm (hunk 3 above)
Comment (MacBlaze1) [reply code/game/objects/machinery/mortar.dm]: using .Copy() now

File: code/game/objects/machinery/mortar.dm (hunk 4 above)
Comment (MacBlaze1) [reply code/game/objects/machinery/mortar.dm]: renamed var

**Comment by BNGP0** [comment]
Can you also fix this bug that makes the shell disappear when you try to pick up a loaded mortar?
And also make it possible to fire mortar by ctrl or alt clicking it for players who have right mouse button disabled in prefs

File: code/game/objects/machinery/mortar.dm [hunk 8]
```diff
@@ -71,10 +71,17 @@
 
 	// used for keeping track of different mortars and their types for cams
 	var/static/list/id_by_type = list()
+	/// list of linked binoculars to the structure of the mortar, used for continuity to item
+	var/list/linked_struct_binoculars = list()
```
Comment (TiviPlus) [inline code/game/objects/machinery/mortar.dm:75]:
```suggestion
	/// list of linked binoculars to the structure of the mortar, used for continuity to item
	var/list/linked_struct_binoculars
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Rest looks ok from skimming but might not have time to review again

**Pull request merged**